### PR TITLE
improve asyncchannel usage and thread safety guarantees

### PR DESCRIPTION
**Pull request by NagyZoltanPeter**
GM/GE @arnetheduck !

As I started my multi-thread brokers implementation based on AsyncChannel thread IPC I faced with some use cases that shown that without thread safety guaranties in AsyncChannels the only option I have, in order to avoid possible crash scenarios, is to intentionally leak the channel.

With this PR I try to solve such issues as using the channel after close (which can happen also in race conditions), or such UB's like use without open.

Extended tests with edge cases and stress test conditions.

I hope you will find it useful and valuable add-on.

### CODE REVIEW

File: asyncchannels.nim [hunk 1]
```diff
@@ -1,20 +1,31 @@
...
 
+# Packed state layout: bit 0 means "open", bit 1 means "closing", and the
+# remaining high bits store the number of in-flight send/recv operations.
+# Each active operation adds/subtracts `asyncChannelRefUnit`, leaving the flag
+# bits untouched while `close` waits for the count to drain to zero.
+const
+  asyncChannelOpenBit = 1'u
+  asyncChannelClosingBit = 2'u
+  asyncChannelRefShift = 2
+  asyncChannelRefUnit = 1'u shl asyncChannelRefShift
```
Comment (gmelodie) [inline asyncchannels.nim:17]: Can this be an `enum`?

File: asyncchannels.nim [hunk 2]
```diff
@@ -68,6 +79,27 @@ type AsyncChannel*[TMsg] = object
     #      for some reason, `Channel` holds a lock while performing the copy
 
   sig: ThreadSignalPtr
+  state: Atomic[uint]
+
+template activeOps(state: uint): uint =
```
Comment (gmelodie) [inline asyncchannels.nim:84]: nit
```suggestion
template activeOperations(state: uint): uint =
```

Comment (gmelodie) [reply asyncchannels.nim:84]: same for similar names, if you decide to go with this

File: asyncchannels.nim [hunk 3]
```diff
@@ -68,6 +79,27 @@ type AsyncChannel*[TMsg] = object
...
+  state shr asyncChannelRefShift
+
+proc misuse(op: static string) {.noreturn, noinline.} =
+  raiseAssert "AsyncChannel." & op & " on a channel that is not open"
+
+proc beginOp[T](tc: ptr AsyncChannel[T]): bool =
+  while true:
+    var state = tc.state.load(moAcquire)
+
+    if (state and asyncChannelOpenBit) == 0 or (state and asyncChannelClosingBit) != 0:
```
Comment (gmelodie) [inline asyncchannels.nim:94]: Maybe have a utility for this 
```suggestion
    if not state.has(asyncChannelOpenBit) or state.has(asyncChannelClosingBit):
```

File: asyncchannels.nim [hunk 4]
```diff
@@ -68,6 +79,27 @@ type AsyncChannel*[TMsg] = object
...
+      return false
+
+    let desired = state + asyncChannelRefUnit
+    if tc.state.compareExchange(state, desired, moAcquireRelease, moAcquire):
+      return true
```
Comment (gmelodie) [inline asyncchannels.nim:99]: is this the same of?
```suggestion
    return tc.state.compareExchange(state, desired, moAcquireRelease, moAcquire):
```

File: asyncchannels.nim [hunk 5]
```diff
@@ -81,21 +113,41 @@ template send2*(c: var Channel, msg: auto) =
...
-  if tc.sig.isNil:
-    return
+  ## Release the resources used by the chanel.
+  ##
+  ## Once `close` starts, new operations are rejected and blocked receivers are
+  ## woken up until all in-flight operations have left the shared state.
+  while true:
+    var state = tc.state.load(moAcquire)
+
+    if (state and asyncChannelOpenBit) == 0 or (state and asyncChannelClosingBit) != 0:
```
Comment (gmelodie) [inline asyncchannels.nim:137]: same as above

File: asyncchannels.nim [hunk 6]
```diff
@@ -81,21 +113,41 @@ template send2*(c: var Channel, msg: auto) =
...
+    if tc.state.compareExchange(state, desired, moAcquireRelease, moAcquire):
+      break
+
+  while true:
+    discard tc.sig.fireSync()
+
+    if tc.state.load(moAcquire).activeOps == 0:
+      break
+
+    sleep(1)
```
Comment (gmelodie) [inline asyncchannels.nim:150]: make this configurable?

File: asyncchannels.nim [hunk 7]
```diff
@@ -81,21 +113,41 @@ template send2*(c: var Channel, msg: auto) =
...
+  if tc.state.load(moAcquire) != 0:
+    return err("AsyncChannel.open on a channel that is already open or closed")
+
   tc.sig = ?ThreadSignalPtr.new()
   # It should not matter that the channel is created in shared memory (since it
   # does not interact with the GC at all) but just to be safe, let's do like the
   # documentation suggests and allocate it like so.
   tc.chan = createShared(typeof(tc.chan[]))
   tc.chan[].open(0)
+  tc.state.store(asyncChannelOpenBit, moRelease)
```
Comment (arnetheduck) [inline asyncchannels.nim:125]: race between loading above and storing here, if two threads call open

File: asyncchannels.nim [hunk 8]
```diff
@@ -81,21 +113,41 @@ template send2*(c: var Channel, msg: auto) =
...
+    var state = tc.state.load(moAcquire)
+
+    if (state and asyncChannelOpenBit) == 0 or (state and asyncChannelClosingBit) != 0:
+      return
+
+    let desired = state or asyncChannelClosingBit
+    if tc.state.compareExchange(state, desired, moAcquireRelease, moAcquire):
+      break
+
+  while true:
```
Comment (arnetheduck) [inline asyncchannels.nim:144]: If there is no reader, this will block forever - this is generally not desireable as it leads to hung applications where it's hard to diagnose what's going on - graceful closing was left to the user of the channel in the initial design in part because of this problem, where the consumer might no longer be around - specially due to the unbounded nature of this channel.

There are a few solutions to this problem - they generally revolve around disconnecting the close operation from what the readers are doing by making close non-blocking and making reading from a closed channel a regular operation that either returns an non-panic error (rust) or a default value (golang) in a non-blocking way.

Comment (NagyZoltanPeter) [reply asyncchannels.nim:144]: @arnetheduck, @gmelodie: Thank you for the review and comments.

I will close this PR without merging.
The reason is, that I ran into severe issues using AsyncChannel[T] or even Channel[T] in my nim-brokers.
All these connected to tear down and only way I could handle those is just leaking them away.
- I could not drain the channel properly.
- somewhat related to the use of channels - I encountered race condition in nim allocators under heavy threading conditions.
- The used ThreadSignalPtr uses 2 file descriptor and thus it did not scale with my broker concept.

Due to these I already abandoned the use of it in my project and refactored into a better match.

**Comment by NagyZoltanPeter** [comment]
Closed as does not solve the issue properly.

File: asyncchannels.nim (hunk 8 above)
Comment (arnetheduck) [reply asyncchannels.nim:144]: > I could not drain the channel properly.

Draining is typically done by posting a sentinel value that signals end-of-stream, then when all consumers are done, signalling to the "stopper" that it can shut down the channel - the problem is similar to closing `go` channels, though they have an explicit "end-of-stream" features while here we need to dedicate a value for it (ie T needs to have a "magic" end-of-stream marker) - possibly, with an approach like these flags, this eos marker could be part of the stream which arguably is easier to reason about.

> I encountered race condition in nim allocators

How did you identify this / is it documented somewhere?

> The used ThreadSignalPtr uses 2 file descriptor and thus it did not scale with my broker concept.

Yes, this is somewhat annoying - there's an alternative design where we use a socketpair / equivalent instead shared between all channels and write the pointer to the stream - it has the advantage that it doesn't need per-channel signalling - the tricky part is to wake the event loop in a platform-agnostic way. If you can think of any other options, those are welcome too.
